saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/serviceprovider/provider/InMemorySaml2RelyingPartyRepository.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.saml2.serviceprovider.provider;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import static java.util.Arrays.asList;
+import static org.springframework.util.Assert.notEmpty;
+import static org.springframework.util.Assert.notNull;
+
+public class InMemorySaml2RelyingPartyRepository
+		implements Saml2RelyingPartyRepository, Iterable<Saml2RelyingPartyRegistration> {
+
+	private final Map<String, Saml2RelyingPartyRegistration> byId;
+	private final Map<String, Saml2RelyingPartyRegistration> byAlias;
+	private final Saml2RelyingPartyRegistration defaultRegistration;
+
+
+	public InMemorySaml2RelyingPartyRepository(Saml2RelyingPartyRegistration... registrations) {
+		this(asList(registrations));
+	}
+
+	public InMemorySaml2RelyingPartyRepository(Collection<Saml2RelyingPartyRegistration> registrations) {
+		notEmpty(registrations, "registrations cannot be empty");
+		byId = createMappingToIdentityProvider(registrations, Saml2RelyingPartyRegistration::getRemoteIdpEntityId);
+		byAlias = createMappingToIdentityProvider(registrations, Saml2RelyingPartyRegistration::getAlias);
+		defaultRegistration = registrations.iterator().next();
+	}
+
+	private static Map<String, Saml2RelyingPartyRegistration> createMappingToIdentityProvider(
+			Collection<Saml2RelyingPartyRegistration> idps,
+			Function<Saml2RelyingPartyRegistration,
+					String> mapper
+	) {
+		LinkedHashMap<String, Saml2RelyingPartyRegistration> result = new LinkedHashMap<>();
+		for (Saml2RelyingPartyRegistration idp : idps) {
+			notNull(idp, "relying party collection cannot contain null values");
+			String key = mapper.apply(idp);
+			notNull(idp, "relying party key may not be null");
+			Assert.isNull(result.get(key), () -> "relying party duplicate key:" + key);
+			result.put(key, idp);
+		}
+		return result;
+	}
+
+	@Override
+	public Saml2RelyingPartyRegistration findByEntityId(String entityId) {
+		Assert.notNull(entityId, "entityId must not be null");
+		return byId.get(entityId);
+	}
+
+	@Override
+	public Saml2RelyingPartyRegistration findByAlias(String alias) {
+		if (StringUtils.hasText(alias)) {
+			return byAlias.get(alias);
+		}
+		else {
+			return defaultRegistration;
```

Comment:
It would surprise me to get back a default registration when I'd asked for one with the given alias. Now, when I call this, I need to check whether or not the one given to me actually has the alias I asked for.

What am I misunderstanding about aliases such that we'd do this kind of fallback for this and not for `findByEntityId`?

Reply:
if you get back defaultRegistration, it means you gave no alias. there is no _given alias_ in this use case.

Reply:
For example 
```
RelyingPartyRegistration rp = relyingPartyRegistrationRepository.findByRegistrationId(
    this.matcher.matcher(request).getVariables().get("registrationId")
);
```

`matcher(request).getVariables().get("registrationId")` may return null, and the underlying implementation decides what to return.